File: apps/server-py/src/lib/emitter.py
from src.lib.socket import sio
from src.lib.firestore import db
from src.lib.logger import logger
from src.lib.utils import now_iso
from src.lib.socket_events import (
    TaskUpdatedPayload, AgentLogPayload,
    PipelineStagePayload, DeploymentUpdatedPayload,
)


def _room(project_id: str) -> str:
    return f"project:{project_id}"


async def emit_task_updated(payload: TaskUpdatedPayload):
    await sio.emit("task:updated", payload.model_dump(), room=_room(payload.projectId))


async def emit_agent_log(payload: AgentLogPayload):
    # Persist to Firestore for observability
    try:
        db.collection("observabilityLogs").add({
            "level": payload.level.upper(),
            "source": "agent",
            "message": payload.message,
            "projectId": payload.projectId,
            "taskId": payload.taskId,
            "agentType": payload.agentType,
            "agentRunId": payload.agentRunId,
            "createdAt": payload.timestamp or now_iso(),
        })
    except Exception as e:
        logger.warning(f"Failed to persist observability log: {e}")
    
    await sio.emit("agent:log", payload.model_dump(), room=_room(payload.projectId))


async def emit_pipeline_stage(payload: PipelineStagePayload):
    await sio.emit("pipeline:stage", payload.model_dump(), room=_room(payload.projectId))


async def emit_deployment_updated(payload: DeploymentUpdatedPayload):
    await sio.emit("deployment:updated", payload.model_dump(mode="json"), room=_room(payload.projectId))

File: apps/server/src/lib/emitter.ts
import { getIO } from "./socket";
import type {
  TaskUpdatedPayload,
  AgentLogPayload,
  JobProgressPayload,
  PipelineStagePayload,
  DeploymentUpdatedPayload,
} from "./socket.events";

// thin helper so services don't have to call getIO() directly
function toProject(projectId: string) {
  return getIO().to(`project:${projectId}`);
}

export const emitter = {
  taskUpdated(payload: TaskUpdatedPayload) {
    toProject(payload.projectId).emit("task:updated", payload);
  },
  agentLog(payload: AgentLogPayload) {
    toProject(payload.projectId).emit("agent:log", payload);
  },
  jobProgress(payload: JobProgressPayload) {
    toProject(payload.projectId).emit("job:progress", payload);
  },
  pipelineStage(payload: PipelineStagePayload) {
    toProject(payload.projectId).emit("pipeline:stage", payload);
  },
  deploymentUpdated(payload: DeploymentUpdatedPayload) {
    toProject(payload.projectId).emit("deployment:updated", payload);
  },
};
